### PR TITLE
 [IMP] mail: delete time model, add clock and clock watcher models

### PR DESCRIPTION
**Pull request by oomsveta**
The Time model is deleted and replaced with the new Clock model. The differences are that there can be several clocks, whereas Time was a singleton, and that a clock must define the frequency at which it updates.

The new model ClockWatcher is added to model a record that makes use of a clock. This is currently useful for the clocks to set or clear intervals as needed.

### CODE REVIEW

File: addons/mail/static/src/models/time_user/time_user.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/** @odoo-module **/
+
+import { registerModel } from '@mail/model/model_core';
+import { attr, one } from '@mail/model/model_field';
+import { insertAndReplace } from '@mail/model/model_field_command';
+
+registerModel({
+    name: 'TimeUser',
```
Comment (seb-odoo) [inline addons/mail/static/src/models/time_user/time_user.js:8]: name to brainstorm

File: addons/mail/static/src/models/time/time.js [hunk 2]
```diff
@@ -17,7 +18,12 @@ registerModel({
          * @returns {number}
          */
         _computeEveryMinuteIntervalId() {
-            return setInterval(this._onEveryMinuteTimeout, 60 * 1000);
+            if (this.timeUsers.some(tu => tu.frequency === 'minute')) {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/time/time.js:21]: Ok to split in various fields to be able to just use length and not iterate

File: addons/mail/static/tests/time_model_tests.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/** @odoo-module **/
+
+import { insertAndReplace } from '@mail/model/model_field_command';
+import { afterEach, beforeEach, start } from '@mail/utils/test_utils';
+
+QUnit.module('mail', {}, function () {
+QUnit.module('Time Model', {
```
Comment (seb-odoo) [inline addons/mail/static/tests/time_model_tests.js:7]: file name instead

File: addons/mail/static/tests/time_model_tests.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            this.widget = widget;
+            return res;
+        };
+    },
+    afterEach() {
+        afterEach(this);
+    },
+});
+
+QUnit.test('everyMinuteIntervalId field on Time model is correctly computed', async function (assert) {
```
Comment (seb-odoo) [inline addons/mail/static/tests/time_model_tests.js:25]: this should probably be 2 or 3 tests with single assert: one that it is empty (by default), one that it is not empty when used,
and maybe one more for the transition from used to not used

File: addons/mail/static/src/models/time_user/time_user.js (hunk 1 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/time_user/time_user.js:8]: Name's strange at first, but when I see what it models I think it's fine.
I think we should take the habit of adding small docstring above all our models.
Here the name feels off because we don't know how to read it.
But it would be totally fine if it says in plain English _"This models records that rely on Time intervals"_: we know it focuses on `Record` <=> `Time` relationship, and it's about `interval` in `Time`, so we get that we should read it as `User of Time`.

File: addons/mail/static/tests/time_model_tests.js (hunk 3 above)
Comment (oomsveta) [reply addons/mail/static/tests/time_model_tests.js:7]: time_model_tests then?

File: addons/mail/static/src/models/activity_view/activity_view.js [hunk 5]
```diff
@@ -66,6 +66,15 @@ registerModel({
             }
             return _.str.sprintf(this.env._t("for %s"), this.activity.assignee.nameOrDisplayName);
         },
+        /**
+         * @private
+         * @returns {ClockWatcher}
+         */
+        _computeClockWatcher() {
+            return insertAndReplace({
+                clock: link(this.messaging.minuteClock),
```
Comment (alexkuhn) [inline addons/mail/static/src/models/activity_view/activity_view.js:75]: prefer `replace`

File: addons/mail/static/src/models/clock/clock.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    identifyingFields: ['messaging', 'frequency'],
+    lifecycleHooks: {
+        _willDelete() {
+            this.messaging.browser.clearInterval(this.intervalId);
+        },
+    },
+    recordMethods: {
+        /**
+         * @private
+         * @returns {number}
```
Comment (alexkuhn) [inline addons/mail/static/src/models/clock/clock.js:18]: `{integer|FieldCommand}`

File: addons/mail/static/src/models/clock/clock.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                return this.messaging.browser.setInterval(this._onTimeout, this.frequency);
+            } else {
+                this.messaging.browser.clearInterval(this.intervalId);
+                return clear();
+            }
+        },
+        /**
+         * @private
+         */
+        _onTimeout() {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/clock/clock.js:31]: `_onInterval`

Comment (alexkuhn) [reply addons/mail/static/src/models/clock/clock.js:31]: Also this should `{ interval: clear() }`, just to make sure (and clean) `Clock/interval` does not refer to a running interval.

We won't need manual `clear()` if we define our `Interval` models.

File: addons/mail/static/tests/clock_model_tests.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    );
+});
+
+QUnit.test('Removing all watchers of a clock clear the interval.', async function (assert) {
+    assert.expect(1);
+
+    await this.start();
+    const clock = this.messaging.models['Clock'].insert({ frequency: 60 * 1000 });
+    const watcher = this.messaging.models['QUnitTest'].insert({
+        clockWatcher: insertAndReplace({ clock: link(clock) }),
```
Comment (alexkuhn) [inline addons/mail/static/tests/clock_model_tests.js:60]: prefer `replace` over `link`

File: addons/mail/static/tests/clock_model_tests.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    });
+
+    watcher.delete();
+    assert.strictEqual(
+        clock.intervalId,
+        undefined,
+        "the clock must no longer have an interval after all its watchers have been deleted."
+    );
+});
```
Comment (alexkuhn) [inline addons/mail/static/tests/clock_model_tests.js:69]: Name of test is broad, and scenario is quite specific: here it's just tested with single watched and it's asserting no interval when the watched is deleted.

It assumes single watched deletion => single watcher removed => no more watchers so no interval.

File: addons/mail/static/src/models/messaging/messaging.js [hunk 10]
```diff
@@ -340,6 +340,14 @@ registerModel({
...
             readonly: true,
         }),
+        minuteClock: one('Clock', {
+            default: insertAndReplace({
+                frequency: 60 * 1000,
+            }),
+            isCausal: true,
+            readonly: true,
+            required: true,
+        }),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging/messaging.js:350]: Not really needed as a field on messaging. Could just use `insertAndReplace({ frequency: 60 * 1000 }),` where it is needed, and it will be created if missing, and re-used otherwise.

File: addons/mail/static/src/models/clock_watcher/clock_watcher.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import { one } from '@mail/model/model_field';
+
+/**
+ * Models a record that makes use of a clock.
+ */
+registerModel({
+    name: 'ClockWatcher',
+    identifyingFields: [['activityView']],
+    fields: {
+        activityView: one('ActivityView', {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/clock_watcher/clock_watcher.js:13]: I would suggest `activityViewOwner` as name to follow our guideline to naming "parent" identifying fields.

File: addons/mail/static/src/models/activity_view/activity_view.js [hunk 12]
```diff
@@ -66,6 +66,17 @@ registerModel({
...
+         * @private
+         * @returns {ClockWatcher}
+         */
+        _computeClockWatcher() {
+            return insertAndReplace({
+                clock: insertAndReplace({
+                    frequency: 60 * 1000,
+                }),
+            });
+        },
```
Comment (seb-odoo) [inline addons/mail/static/src/models/activity_view/activity_view.js:79]: Minor details:

The method returns a command not a ClockWatcher.

Also this could probably be a default instead as all values are static.

File: addons/mail/static/src/models/clock/clock.js [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+         */
+        _onInterval() {
+            this.update({ date: new Date() });
+        },
+    },
+    fields: {
+        /**
+         * A Date object set at the current date, updated at every tick.
+         */
+        date: attr({
```
Comment (seb-odoo) [inline addons/mail/static/src/models/clock/clock.js:45]: note: confusing name

should be browser.date/web.date?

but it's actually a date + time (even the browser one is badly named)

**Review by alexkuhn — APPROVED** [review]

File: addons/mail/static/tests/clock_model_tests.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    const watcher = this.messaging.models['ClockWatcher'].insert({
+        qunitTestOwner: insertAndReplace(),
+        clock: insertAndReplace({ frequency: 180 * 1000 }),
+    });
+
+    watcher.delete();
+    assert.strictEqual(
+        this.messaging.models['Clock'].all().length,
+        0,
+        "deleting all the watchers of a clock should result in the deletion of the clock itself."
```
Comment (seb-odoo) [inline addons/mail/static/tests/clock_model_tests.js:38]: single assert test should ideally have assert comment equals test title (not sure which one is best in this case, maybe a mix of both)

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
